File: modules/onno-telegram-login/android/src/main/java/expo/modules/onnotelegramlogin/OnnoTelegramLoginModule.kt
package expo.modules.onnotelegramlogin

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

// Native bridge for "Login with Telegram" on Android.
//
// Wraps Telegram's official login SDK (TelegramMessenger/telegram-login-android). The SDK launches
// the Telegram app when installed and otherwise falls back to a Chrome Custom Tab; on success it
// returns an OIDC ID token (a JWT) that the JS layer POSTs to /api/auth/telegram/native.
//
// Promise contract (consumed by src/auth/telegramLogin.ts):
//   resolve(mapOf("idToken" to String, "viaWebFallback" to Boolean))
//   reject("ERR_TELEGRAM_CANCELLED", …)   — user dismissed the flow
//   reject("ERR_TELEGRAM_UNAVAILABLE", …) — SDK not wired into this build
//   reject("ERR_TELEGRAM_FAILED", …)      — anything else
//
// The callback redirect (app link / custom scheme) intent-filter and the <queries> Telegram-package
// allow-list are configured by the Expo config plugin: plugins/withTelegramLogin.js (+ this module's
// AndroidManifest.xml).

class OnnoTelegramLoginModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("OnnoTelegramLogin")

    AsyncFunction("login") { options: Map<String, Any?>, promise: Promise ->
      val nonce = options["nonce"] as? String
      startLogin(nonce, promise)
    }
  }

  private fun startLogin(nonce: String?, promise: Promise) {
    val activity = appContext.currentActivity
    if (activity == null) {
      promise.reject("ERR_TELEGRAM_FAILED", "No foreground activity to present Telegram login.", null)
      return
    }

    // ────────────────────────────────────────────────────────────────────────────────
    // TODO(native): wire the official Telegram login SDK here. Sketch of the expected
    // shape (confirm the exact symbols against the upstream README before building):
    //
    //   val request = TelegramLoginRequest.Builder(botId = <yourBotId>)
    //     .scope(Scope.IDENTITY)
    //     .nonce(nonce)
    //     .callbackUrl("onno-telegram://telegram-login")
    //     .build()
    //   TelegramLogin.authenticate(activity, request, object : TelegramLoginCallback {
    //     override fun onSuccess(idToken: String, viaWebFallback: Boolean) {
    //       promise.resolve(mapOf("idToken" to idToken, "viaWebFallback" to viaWebFallback))
    //     }
    //     override fun onCancelled() {
    //       promise.reject("ERR_TELEGRAM_CANCELLED", "The user cancelled Telegram sign-in.", null)
    //     }
    //     override fun onError(error: Throwable) {
    //       promise.reject("ERR_TELEGRAM_FAILED", error.message, error)
    //     }
    //   })
    //   return
    // ────────────────────────────────────────────────────────────────────────────────

    // Until the SDK above is linked, report "unavailable" so the JS layer falls back to the
    // server's web SSO flow instead of dead-ending on the button.
    @Suppress("UNUSED_EXPRESSION") nonce
    promise.reject(
      "ERR_TELEGRAM_UNAVAILABLE",
      "Telegram login SDK is not yet wired into this build. See modules/onno-telegram-login/README.md.",
      null
    )
  }
}

File: modules/onno-telegram-login/ios/OnnoTelegramLoginModule.swift
import ExpoModulesCore

#if canImport(TelegramLogin)
import TelegramLogin
#endif

// Native bridge for "Login with Telegram" on iOS.
//
// Wraps Telegram's official login SDK (TelegramMessenger/telegram-login-ios, added via Swift Package
// Manager). The SDK opens the Telegram app when installed and otherwise presents a secure
// ASWebAuthenticationSession; on success it returns an OIDC ID token (a JWT) that the JS layer POSTs
// to /api/auth/telegram/native.
//
// Config (clientId / redirectUri / scopes) is written into Info.plist by the Expo config plugin
// (plugins/withTelegramLogin.js) so nothing is hardcoded here. The redirect callback is delivered by
// OnnoTelegramLoginAppDelegate (registered as an Expo AppDelegate subscriber) which forwards the URL
// to `TelegramLogin.handle(_:)`.
//
// Promise contract (consumed by src/auth/telegramLogin.ts):
//   resolve(["idToken": String, "viaWebFallback": Bool])
//   reject("ERR_TELEGRAM_CANCELLED", …)   — user dismissed the sheet
//   reject("ERR_TELEGRAM_UNAVAILABLE", …) — SDK not linked / not configured
//   reject("ERR_TELEGRAM_FAILED", …)      — anything else
//
// The `#if canImport(TelegramLogin)` guard keeps the module compiling even before the SPM package is
// added (it then reports "unavailable", so the app falls back to the server's web SSO flow).

public class OnnoTelegramLoginModule: Module {
  public func definition() -> ModuleDefinition {
    Name("OnnoTelegramLogin")

    AsyncFunction("login") { (options: [String: Any?], promise: Promise) in
      let nonce = options["nonce"] as? String
      DispatchQueue.main.async {
        OnnoTelegramLoginModule.startLogin(nonce: nonce, promise: promise)
      }
    }
  }

  static func startLogin(nonce: String?, promise: Promise) {
    #if canImport(TelegramLogin)
    guard let cfg = TelegramLoginConfig.fromInfoPlist() else {
      promise.reject("ERR_TELEGRAM_UNAVAILABLE", "Telegram login is not configured. Add the plugin options in app.json (appId/clientId).")
      return
    }
    // `nonce` is reserved for replay protection; the current SDK doesn't take one, so it's bound by
    // the server (/api/auth/telegram/native/begin) rather than threaded through the SDK here.
    _ = nonce

    TelegramLoginConfig.ensureConfigured(cfg)

    TelegramLogin.login { result in
      switch result {
      case .success(let loginData):
        promise.resolve(["idToken": loginData.idToken, "viaWebFallback": false])
      case .failure(let error):
        switch error {
        case .cancelled:
          promise.reject("ERR_TELEGRAM_CANCELLED", "The user cancelled Telegram sign-in.")
        default:
          promise.reject("ERR_TELEGRAM_FAILED", error.localizedDescription)
        }
      }
    }
    #else
    _ = nonce
    promise.reject(
      "ERR_TELEGRAM_UNAVAILABLE",
      "Telegram login SDK is not linked in this build. Add it via Swift Package Manager — see modules/onno-telegram-login/README.md."
    )
    #endif
  }
}

#if canImport(TelegramLogin)
/// Reads the Info.plist values the config plugin writes, and configures the SDK once.
struct TelegramLoginConfig {
  let clientId: String
  let redirectUri: String
  let scopes: [String]

  private static var configured = false

  static func fromInfoPlist() -> TelegramLoginConfig? {
    let info = Bundle.main.infoDictionary
    guard
      let clientId = info?["TelegramLoginClientId"] as? String, !clientId.isEmpty,
      let redirectUri = info?["TelegramLoginRedirectUri"] as? String, !redirectUri.isEmpty
    else {
      return nil
    }
    let scopes = (info?["TelegramLoginScopes"] as? [String]) ?? ["profile"]
    return TelegramLoginConfig(clientId: clientId, redirectUri: redirectUri, scopes: scopes)
  }

  static func ensureConfigured(_ cfg: TelegramLoginConfig) {
    guard !configured else { return }
    TelegramLogin.configure(clientId: cfg.clientId, redirectUri: cfg.redirectUri, scopes: cfg.scopes)
    configured = true
  }

  /// Configure from Info.plist if possible (used at app launch by the AppDelegate subscriber).
  static func configureFromInfoPlistIfPossible() {
    if let cfg = fromInfoPlist() {
      ensureConfigured(cfg)
    }
  }
}
#endif
